### ISSUE TITLE
Figure out javascript-less forms

### ISSUE THREAD

**Issue by Mstrodl** [sveltejs/sapper#634]
We should ideally be able to use sapper to make a page that will work with or without javascript. For example, a signup page should be able to use a `<form>` should function properly:
```html
<!-- routes/signup/index.svelte -->
<form action="/signup" method="post">
    <input type="email" name="email" bind:value="email" />
    {#if emailError}
        <span class="error">{emailError}</span>
    {/if}
    <input type="password" name="password" bind:value="password" />
    <input type="submit" name="signup">
</form>
```

```js
// routes/signup/index.js
export function post(req, res) {
    // We fall through to render 'routes/signup/index.svelte'
    res.render({
        emailError: "This email is already taken!",
        email: req.body.email,
        password: req.body.password
    });
}
```
The data `render()` gets should be merged with whatever data the page's component has. It could also get passed as another option on `preload()`, however that might not be the best solution as it means users with JS would need to re-fetch things.

As far as the internal bits:
 - If the user has JS, we could intercept the form submission (`on:submit`) and use `fetch` to POST data to '/signup' with an 'Accept: application/json' header
 - If the user does not have JS, '/signup' would not have an 'Accept: application/json' and we would thus serve a new rendered page

In other words:
```js
res.render = function(data) {
    if(req.headers.accept == "application/json") {
        res.setHeader("Content-Type", "application/json");
        res.writeHead(200);
        res.end(JSON.stringify(data));
    } else {
        const result = PageComponent.render({
            ...data,
            // not 100% sure how preloading works internally so excuse this
            ...preloadProps
        });
        res.setHeader("Content-Type", "text/html");
        res.writeHead(200);
        res.end(buildPage(result));
    }
}
```

**Comment by Mstrodl** [comment]
Actually, ideally we should be able to send the user a redirect to another page too when they do have javascript, so perhaps we would want some sort of format like this: ```json
// 0 - set props
[0, {"emailError":"This email is already taken!"}]
// 1 - redirect
[1,"/login",{"message":"Account created! Log in below!"}]
```
(Note that this is just a super minimal version in the interest of efficiency. It might just make more sense to have an object like `{"action": "redirect", "path": "/login", "data": {"message": "Account created!"}}`. Ideally a redirect would also run the preload() script server-side to minimize the number of requests)

**Comment by Rich-Harris** [comment]
Possibly relevant — just merged https://github.com/sveltejs/sapper/pull/628 which reinstates support for treating a server route as middleware for a page route. It probably doesn't fully solve the issue but it's a necessary first step.
